integration/spark/app/src/test/java/io/openlineage/spark/agent/util/RunEventVerifier.java

```diff
@@ -0,0 +1,283 @@
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import lombok.SneakyThrows;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang.StringUtils;
+import org.mockserver.serialization.ObjectMapperFactory;
+import shaded_package.com.fasterxml.jackson.core.JsonProcessingException;
+import shaded_package.com.fasterxml.jackson.databind.JsonNode;
+import shaded_package.com.fasterxml.jackson.databind.JsonSerializer;
+import shaded_package.com.fasterxml.jackson.databind.ObjectWriter;
+import shaded_package.net.javacrumbs.jsonunit.core.Configuration;
+import shaded_package.net.javacrumbs.jsonunit.core.Option;
+import shaded_package.net.javacrumbs.jsonunit.core.internal.Diff;
+import shaded_package.net.javacrumbs.jsonunit.core.internal.Options;
+import shaded_package.net.javacrumbs.jsonunit.core.listener.DifferenceContext;
+import shaded_package.net.javacrumbs.jsonunit.core.listener.DifferenceListener;
+
+/**
+ * Utility class to verify if expected JSON content of the OpenLineage event is present among the
+ * events emitted.
+ */
+@Slf4j
+public class RunEventVerifier {
+
+  private final List<String> events;
+  private String diffDescription;
+
+  @SneakyThrows
+  private RunEventVerifier(List<String> events) {
+    this.events = events;
+  }
+
+  /**
+   * Path to a file containing in each line a separate JSON OpenLineage event.
+   *
+   * @param eventsFile
+   * @return
+   */
+  @SneakyThrows
+  public static RunEventVerifier of(File eventsFile) {
+    return new RunEventVerifier(Files.readAllLines(eventsFile.toPath()));
+  }
+
+  @SneakyThrows
+  public static RunEventVerifier ofStrings(List<String> events) {
+    return new RunEventVerifier(events);
+  }
+
+  @SneakyThrows
+  public static RunEventVerifier ofRunEvents(List<RunEvent> events) {
+    return new RunEventVerifier(
+        events.stream().map(e -> OpenLineageClientUtils.toJson(e)).collect(Collectors.toList()));
+  }
+
+  @SneakyThrows
+  public boolean match(File... expectationFiles) {
+    boolean matchesAll = true;
+    for (File expectationFile : expectationFiles) {
+      String expectedJson =
+          new String(Files.readAllBytes(expectationFile.toPath()), StandardCharsets.UTF_8);
+      JsonStringMatcher matcher = new JsonStringMatcher(expectedJson);
+
+      // each event has to be valid for some of the events contained
+      boolean matches = events.stream().filter(e -> matcher.matches(e)).findAny().isPresent();
+
+      if (!matches) {
+        diffDescription =
+            new StringBuilder()
+                .append("Couldn't match event for ")
+                .append(expectationFile.getName())
+                .append(System.lineSeparator())
+                .append(bestEffortDiffDescription(events, expectedJson))
+                .append(System.lineSeparator())
+                .toString();
+      }
+
+      matchesAll = matchesAll && matches;
```

Comment:
you can also look into `Arrays.stream(expectationFiles).allMatch()`

Reply:
looks sweet but docs says `This is a short-circuiting terminal operation.`, which will not collect full `diffDescription` in case of multiple files not being matched.